src/storage/txn/commands/flashback_to_version.rs

```diff
@@ -84,29 +83,43 @@ impl<S: Snapshot, L: LockManager> WriteCommand<S, L> for FlashbackToVersion {
                 rows += 1;
             }
         }
-        // To flashback the `CF_WRITE`, we need to delete all write records whose
-        // `commit_ts` is greater than the specified version, and if it's not a
-        // short-value `WriteType::Put`, we need to delete the actual data from
-        // `CF_DEFAULT` as well.
-        for (key, write) in self.key_writes {
+        // To flashback the `CF_WRITE` and `CF_DEFAULT`, we need to write a new MVCC
+        // record for each key in `self.keys` with its old value at `self.version`,
+        // specifically, the flashback will have the following behavior:
+        //   - If a key doesn't exist at `self.version`, it will be put a
+        //     `WriteType::Delete`.
+        //   - If a key exists at `self.version`, it will be put the exact same record
+        //     in `CF_WRITE` and `CF_DEFAULT` if needed with `self.commit_ts` and
+        //     `self.start_ts`.
+        for key in self.keys {
             if txn.write_size() >= MAX_TXN_WRITE_SIZE {
                 next_write_key = Some(key);
                 break;
             }
-            let encoded_key = key.clone().truncate_ts()?;
-            let commit_ts = key.decode_ts()?;
-            txn.delete_write(encoded_key.clone(), commit_ts);
-            rows += 1;
-            // If the short value is none and it's a `WriteType::Put`, we should delete the
-            // corresponding key from `CF_DEFAULT` as well.
-            if write.short_value.is_none() && write.write_type == WriteType::Put {
-                txn.delete_value(encoded_key, write.start_ts);
+            let old_write = reader.seek_write(&key, self.version)?;
+            // If the old write doesn't exist, we should put a `WriteType::Delete` record to
+            // delete the current key.
+            if old_write.is_none() {
+                let new_write = Write::new(WriteType::Delete, self.start_ts, None);
+                txn.put_write(key.clone(), self.commit_ts, new_write.as_ref().to_bytes());
                 rows += 1;
+                continue;
             }
+            let old_write = old_write.unwrap().1;
+            // If it's not a short value and it's a `WriteType::Put`, we should put the old
+            // value in `CF_DEFAULT` with `self.start_ts` as well.
+            if old_write.short_value.is_none() && old_write.write_type == WriteType::Put {
+                if let Some(old_value) = reader.get(&key, self.version)? {
+                    txn.put_value(key.clone(), self.start_ts, old_value);
+                    rows += 1;
+                }
+            }
+            let new_write = Write::new(old_write.write_type, self.start_ts, old_write.short_value);
+            txn.put_write(key.clone(), self.commit_ts, new_write.as_ref().to_bytes());
```

Comment:
When read meets write record with `WriteType::Lock` or `WriteType::Rollback`, it'll continue to next value. In this case, you should find the latest visible value and write it unless a `WriteType::Lock` or `WriteType::Rollback` will fallback to the wrong result.